### PR TITLE
[RFC] osd: ec partial stripe reads

### PR DESCRIPTION
**Pull request by thinkercui**
If a read is limited at one stripe and all the original data
chunks are available. We will only read the chunks that contain
range <off, len> to reduce the read latency and bandwidth.

If the read is fast read or not all the original data chunks are
available we will do a normal read.

This idea comes from http://permalink.gmane.org/gmane.comp.file-systems.ceph.devel/23872.

Signed-off-by: Xiaofei Cui <cuixiaofei@sangfor.com.cn>

### CODE REVIEW

**Comment by thinkercui** [comment]
retest it please

**Comment by thinkercui** [comment]
retest this please

**Comment by thinkercui** [comment]
retest this please

**Comment by thinkercui** [comment]
retest this please

**Comment by gregsfortytwo** [comment]
This doesn't seem right:
* it's open-coding the striping strategy without any reference to the erasure code plugin in use?
* don't we already read only the necessary shards by asking the EC plugin what the minimum-to-decode is?

Maybe we still do a full-shard read and I've made up that we're efficient about that, but it seems like extending the interfaces around minimum-to-decode is the way to go here.

**Comment by thinkercui** [comment]
Thanks for your review.  @gregsfortytwo 

- This is for all the erasure code plugin but not for some one.
- We need to tell what shards we want before we ask the EC plugin what the minimum-to-decode is.
The less shards we want the less shards we need to read. The main point of this commit is to reduce
want to read shards.

Do I get the point what you mean? Thanks.

**Comment by gregsfortytwo** [comment]
Maybe you'd better write some docs explaining the strategy then. When I skim this, it looks like you're trying to map from logical data offsets to which chunks to read, in the OSD code, without reference to the EC plugin in use. So I think you're just assuming that the EC plugin's data placement is systematic and is a simple linear stream across the objects? That is frequently not the case for more exotic EC codes (some of which we have in-tree).

So it's not that we need to tell the EC plugin what shards we want to read; we have to tell the EC plugin what data offsets we want to read, and let it tell us the minimum number of reads to get that data.

File: src/osd/ECBackend.cc [hunk 1]
```diff
@@ -2295,11 +2323,51 @@ void ECBackend::objects_read_and_reconstruct(
   }
 
   map<hobject_t, set<int>> obj_want_to_read;
-  set<int> want_to_read;
-  get_want_to_read_shards(&want_to_read);
```
Comment (jdurgin) [inline src/osd/ECBackend.cc]: this seems like a much larger change than necessary... here's all it took to make recovery read the minimum data required: https://github.com/ceph/ceph/pull/21273/commits/468ad4b41010488c8d48ef65ccbebfdb4270690f

**Comment by thinkercui** [comment]
You are right! I assume that the EC plugin's data placement is systematic and is a simple linear stream across the objects. First, all the EC plugins of ceph are systematic right now. Second, I think the chunk_mapping can make sure the data is a simple linear stream across the objects. In addition to make the code more general, I have made another commit to restrict the partial read to systematic EC plugins.

We really can tell the EC plugin what data offsets we want to read, and let it tell us the minimum number of reads to get that data. But, I think a EC plugin should better only deal with the stripe, but not the whole object, left the work to osd maybe better.

Is there any other suggestion? Thanks. @gregsfortytwo

File: src/osd/ECBackend.cc (hunk 1 above)
Comment (thinkercui) [reply src/osd/ECBackend.cc]: This modification only relates the read from client. The read of recovery is left as it is. So, don't worry about [468ad4b](https://github.com/ceph/ceph/commit/468ad4b41010488c8d48ef65ccbebfdb4270690f)

Comment (jdurgin) [reply src/osd/ECBackend.cc]: I'm suggesting that the client read path can be similarly simple as the recovery read path. There should not be redundancies with the EC plugin or the recovery code.

**Comment by jdurgin** [comment]
Closing due to unaddressed comments, please reopen if you're still working on this

**Pull request closed without merging**